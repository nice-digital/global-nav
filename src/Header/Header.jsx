import React, { Component } from "react";
import PropTypes from "prop-types";
import LogoIcon from "@nice-digital/icons/lib/LogoFull";

import {
	defaultEventCategory,
	headerClickEventAction,
	trackEvent,
} from "./../tracker";
import OldIEMessage from "./OldIEMessage";
import CoronaMessage from "./CoronaMessage";
import Nav from "./Nav";
import Search from "./Search";
import Account from "./Account";
import SkipLink from "./SkipLink";

import styles from "./Header.module.scss";
import { HeaderContextProvider, HeaderContext } from "./context/HeaderContext";

export class Header extends Component {
	constructor(props) {
		super(props);

		this.state = {
			isExpanded: false,
			isLoggedIn: false,
			accountsData: null,
			scrimIsActive: false,
		};

		this.handleMobileMenuBtnClick = this.handleMobileMenuBtnClick.bind(this);
		this.handleLoginStatusChecked = this.handleLoginStatusChecked.bind(this);
		this.handleLogoClick = this.handleLogoClick.bind(this);
	}

	componentDidMount() {
		if (!document.getElementById(this.props.skipLinkId)) {
			const firstH1OnPage = document.getElementsByTagName("h1")[0];
			firstH1OnPage
				? firstH1OnPage.setAttribute("id", this.props.skipLinkId)
				: console.warn(
						`Global nav "skip to link" can't find a H1 tag or an element with the ID of "${this.props.skipLinkId}"`
				  );
		}
	}

	handleMobileMenuBtnClick() {
		trackEvent(defaultEventCategory, headerClickEventAction, "Menu");

		this.setState((prevState) => ({
			isExpanded: !prevState.isExpanded,
		}));
	}

	handleLogoClick(e) {
		e.preventDefault();

		const href = e.currentTarget.getAttribute("href");

		trackEvent(
			defaultEventCategory,
			headerClickEventAction,
			"Logo",
			null,
			href,
			function () {
				window.location.href = href;
			}
		);
	}

	handleLoginStatusChecked(accountsData) {
		if (accountsData.display_name) {
			this.setState({ isLoggedIn: true, accountsData: accountsData });
		} else {
			this.setState({ isLoggedIn: false, accountsData: accountsData });
		}
	}

	getCallbackFunction(funcOrFuncName) {
		if (typeof funcOrFuncName === "function") {
			return funcOrFuncName;
		}

		if (typeof window[funcOrFuncName] === "function") {
			return window[funcOrFuncName];
		}

		return undefined;
	}

	dropdownToggleHandler(idOfOpenDropdown) {
		// If we've got an onDropdownOpen or onDropdownClose prop call the onDropdownOpen or onDropdownClose callback based on idOfOpenDropdown.
		const { onDropdownOpen, onDropdownClose } = this.props;

		if (onDropdownOpen && idOfOpenDropdown !== null) {
			if (typeof this.getCallbackFunction(onDropdownOpen) === "function") {
				onDropdownOpen();
			}
		}

		if (onDropdownClose && idOfOpenDropdown == null) {
			if (typeof this.getCallbackFunction(onDropdownClose) === "function") {
				onDropdownClose();
			}
		}
	}

	render() {
		return (
			this.props.enabled !== false && (
				<HeaderContextProvider>
					<HeaderContext.Consumer>
						{({ idOfOpenDropdown }) => {
							this.dropdownToggleHandler(idOfOpenDropdown);
							return (
								<span
									id="scrim"
									className={
										idOfOpenDropdown !== null ? styles.scrim : undefined
									}
									aria-hidden="true"
								/>
							);
						}}
					</HeaderContext.Consumer>

					<div className={styles.header} data-tracking="Global nav">
						<header aria-label="Site header">
							<ul className={styles.a11yLinks} aria-label="Accessibility links">
								<li>
									<SkipLink to={`#${this.props.skipLinkId}`}>
										Skip to content
									</SkipLink>
								</li>
								<li>
									<SkipLink to="https://www.nice.org.uk/accessibility">
										Accessibility help
									</SkipLink>
								</li>
							</ul>
							<div className={styles.container}>
								<a
									href="https://www.nice.org.uk/"
									aria-label="Home"
									className={styles.home}
									onClick={this.handleLogoClick}
									data-tracking="Logo"
								>
									<LogoIcon
										className={styles.icon}
										width={null}
										height={null}
									/>
								</a>
								<div className={styles.wrapper}>
									<div className={styles.search}>
										{this.props.search && (
											<Search
												skipLinkId={this.props.skipLinkId}
												onNavigating={this.props.onNavigating}
												{...this.props.search}
											/>
										)}
									</div>
									<button
										className={styles.mobileMenuBtn}
										id="header-menu-button"
										type="button"
										aria-controls="header-menu"
										aria-expanded={this.state.isExpanded}
										aria-haspopup="menu"
										aria-label={
											this.state.isExpanded
												? "Close site menu"
												: "Expand site menu"
										}
										onClick={this.handleMobileMenuBtnClick}
									>
										{this.state.isExpanded ? "Close" : "Menu"}
									</button>
									{this.props.auth !== false && (
										<div className={styles.account}>
											<Account
												onLoginStatusChecked={this.handleLoginStatusChecked}
												isLoggedIn={this.state.isLoggedIn}
												accountsData={this.state.accountsData}
												{...this.props.auth}
											/>
										</div>
									)}
								</div>
							</div>
							<Nav
								skipLinkId={this.props.skipLinkId}
								service={this.props.service}
								isExpanded={this.state.isExpanded}
								accountsLinks={
									this.state.accountsData && this.state.accountsData.links
								}
								onNavigating={this.props.onNavigating}
								additionalSubMenuItems={this.props.additionalSubMenuItems}
							/>
						</header>
						<CoronaMessage onResize={this.props.onResize} />
						<OldIEMessage />
					</div>
				</HeaderContextProvider>
			)
		);
	}
}

Header.propTypes = {
	service: PropTypes.string,
	skipLinkId: PropTypes.string,
	enabled: PropTypes.bool,
	search: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
	auth: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
	onNavigating: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
	onResize: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
	onDropdownOpen: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
	onDropdownClose: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
	additionalSubMenuItems: PropTypes.arrayOf(PropTypes.object),
};

Header.defaultProps = {
	search: {},
	skipLinkId: "content-start",
	additionalSubMenuItems: [],
};

export default Header;
