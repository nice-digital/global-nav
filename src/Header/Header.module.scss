@import 'header-settings';

.header {
  background: $nds-colour-nice-white;
  border-bottom: rem($nds-spacing-xx-small) solid $nds-colour-nice-teal;
  box-sizing: border-box;
  font-family: unquote($nds-font-family-sans);

  *,
  *:before,
  *:after {
    box-sizing: inherit;
    font-family: inherit;
  }

  @media print {
    border-bottom: none;
  }
}

.container {
  @include nds-container;
  padding: rem($nds-spacing-medium 0 0 0);

  &:after {
    clear: both;
    content: '';
    display: table;
  }
}

.home {
  @include nds-remove-appended-print-link;
  color: $nds-colour-true-black;
  display: inline-block;
  float: left;
  height: rem(30);
  margin-bottom: rem($nds-spacing-medium);

  &:focus {
    @include nds-default-focus-style;
  }

  &:visited {
    color: $nds-colour-true-black;
  }

  .icon {
    height: 100%;
    width: auto;
  }
}

.wrapper {
  @include nds-hide-print;
  display: flex;
  flex-grow: 1;
  float: right;
  justify-content: space-between;
  margin-bottom: rem($nds-spacing-medium);
  width: 50%;

  &:after {
    clear: both;
    content: '';
    display: table;
  }
}

.search {
  flex-grow: 1;
  float: left;
  margin-right: rem($nds-spacing-medium);
  min-width: percentage(2 / 3);
}

.mobileMenuBtn {
  appearance: none;
  background: none;
  border: 1px solid $nds-colour-text;
  float: right;
  font-family: inherit;
  font-size: 110%;
  height: rem($search-height);
  min-width: rem(75); // Stops width changing when button text changes
  white-space: nowrap;

  &:after {
    border-left: rem($nds-spacing-x-small) solid transparent;
    border-right: rem($nds-spacing-x-small) solid transparent;
    border-top: rem($nds-spacing-small) solid $nds-colour-nice-dark-grey;
    content: '';
    display: inline-block;
    height: 0;
    margin-left: rem($nds-spacing-small);
    vertical-align: middle;
    width: 0;
  }

  &[aria-expanded='true']:after {
    border-bottom: rem($nds-spacing-small) solid $nds-colour-text;
    border-top: 0;
  }
}

.account {
  display: none;
}

@include mq($until: md) {
  .home {
    float: none;
  }

  .wrapper {
    float: none;
    width: auto;
  }
}

@include mq($from: md) {
  .header {
    border-bottom: rem($nds-spacing-x-small) solid $nds-colour-body;
  }

  .container {
    display: flex;
    justify-content: space-between;
  }

  .home {
    height: rem($search-height);
    margin-right: rem($nds-spacing-large);
  }

  .scrim {
    background: rgba(0, 0, 0, 70%);
    display: block;
    height: 100vh;
    left: 0;
    margin-top: rem($search-height) + rem($nds-spacing-large);
    position: absolute;
    width: 100vw;
  }

  .wrapper {
    justify-content: flex-end;
  }

  .search {
    margin-right: rem($nds-spacing-large);
  }

  .mobileMenuBtn {
    display: none;
  }

  .account {
    display: block;
  }
}

.a11yLinks {
  @include nds-hide-print;
  list-style: none;
  margin: 0;
  padding: 0;

  li {
    list-style: none;
    margin: 0;
  }
}
