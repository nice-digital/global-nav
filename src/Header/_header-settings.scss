@use '@nice-digital/nds-core/scss/spacing';

$sub-nav-height: spacing.$x-large;
$search-height: 40;
$base-z-index: 99990;
