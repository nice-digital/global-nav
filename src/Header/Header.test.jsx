import React from "react";
import Header from "./Header";
import { render, within, createEvent, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import {
	eventName,
	defaultEventCategory,
	headerClickEventAction,
	eventTimeout,
} from "./../tracker";

describe("Header", () => {
	const defaultProps = {
		service: null,
		enabled: null,
	};

	beforeEach(() => {
		window.dataLayer = [];
	});

	afterAll(() => {
		// Cleanup
		delete window.dataLayer;
	});

	it("Matches snapshot", () => {
		const { container } = render(
			<Header {...defaultProps} onNavigating="onNavigatingHandler" />
		);
		expect(container).toMatchSnapshot();
	});

	describe("Mobile menu button", () => {
		it("Mobile menu button is collapsed by default", () => {
			const { getByRole } = render(<Header {...defaultProps} />);

			const mobileMenuButton = getByRole("button", {
				name: "Expand site menu",
			});

			expect(mobileMenuButton).toHaveAttribute("aria-expanded", "false");
			expect(mobileMenuButton).toHaveTextContent("Menu");
		});

		it("Mobile menu button toggles text and aria-expanded on click", async () => {
			const { getByRole } = render(<Header {...defaultProps} />),
				mobileMenuButton = getByRole("button", {
					name: "Expand site menu",
				}),
				user = userEvent.setup();

			await user.click(mobileMenuButton);

			expect(mobileMenuButton).toHaveAttribute("aria-expanded", "true");
			expect(mobileMenuButton).toHaveTextContent("Close");
			expect(mobileMenuButton).toHaveAccessibleName("Close site menu");
		});

		it("Mobile menu button toggles nav on click", async () => {
			const { getByRole } = render(<Header {...defaultProps} />),
				mobileMenuButton = getByRole("button", {
					name: "Expand site menu",
				}),
				primaryNav = getByRole("navigation", {
					name: "primary navigation",
					hidden: true,
				}),
				navWrapper = primaryNav.parentElement,
				user = userEvent.setup();

			expect(navWrapper).not.toHaveClass("wrapperExpanded");

			await user.click(mobileMenuButton);
			expect(navWrapper).toHaveClass("wrapperExpanded");

			await user.click(mobileMenuButton);
			expect(navWrapper).not.toHaveClass("wrapperExpanded");
		});

		it("should track mobile menu button click", async () => {
			const { getByRole } = render(<Header {...defaultProps} />),
				mobileMenuButton = getByRole("button", {
					name: "Expand site menu",
				}),
				user = userEvent.setup();

			await user.click(mobileMenuButton);

			expect(window.dataLayer).toEqual([
				{
					event: eventName,
					eventCategory: defaultEventCategory,
					eventAction: headerClickEventAction,
					eventLabel: "Menu",
				},
			]);
		});
	});

	describe("Search", () => {
		it("Renders search by default", () => {
			const { getByRole } = render(<Header {...defaultProps} />);
			expect(getByRole("search")).toBeInTheDocument();
		});

		it("Doesn't render search if search is disabled", () => {
			const { queryByRole } = render(
				<Header {...defaultProps} search={false} />
			);
			expect(queryByRole("search")).toBeNull();
		});

		it("should set search form action URL", () => {
			const { getByRole } = render(
				<Header {...defaultProps} search={{ url: "/test" }} />
			);
			expect(getByRole("search")).toHaveAttribute("action", "/test");
		});

		it("should set search input placeholder", () => {
			const { getByRole } = render(
				<Header {...defaultProps} search={{ placeholder: "Toast" }} />
			);
			expect(getByRole("searchbox")).toHaveAttribute("placeholder", "Toast");
		});

		it("should set search input default text", () => {
			const { getByRole } = render(
				<Header {...defaultProps} search={{ query: "bananas" }} />
			);
			expect(getByRole("searchbox")).toHaveValue("bananas");
		});

		it("should create skip link to given target ID", () => {
			const { getByRole } = render(
					<Header {...defaultProps} search={{ skipLinkId: "chickens" }} />
				),
				search = getByRole("search");

			expect(
				within(search).getByRole("link", { name: "Skip to content" })
			).toHaveAttribute("href", "#chickens");
		});
	});

	describe("Nav", () => {
		it("Nav is collapsed by default", async () => {
			const { getByRole } = render(<Header {...defaultProps} />),
				primaryNav = getByRole("navigation", {
					name: "primary navigation",
					hidden: true,
				}),
				navWrapper = primaryNav.parentElement;

			expect(navWrapper).not.toHaveClass("wrapperExpanded");
		});
	});

	describe("Back to top link target", () => {
		it("should contain a back to top scroll target id", () => {
			const { getByRole } = render(<Header {...defaultProps} />),
				siteHeader = getByRole("banner", { name: "Site header" });

			expect(siteHeader.parentElement).toHaveAttribute("id", "top");
		});
	});

	describe("Skip link target", () => {
		it(`If skip link doesn't exist, create a link to the first <h1> in the document`, () => {
			const heading1 = document.createElement("h1");
			document.body.append(heading1);

			render(<Header {...defaultProps} />);

			expect(heading1).toHaveAttribute("id", "content-start");
			heading1.parentNode.removeChild(heading1);
		});

		it("Renders a skip link which targets the custom skipLinkId property", () => {
			const heading1 = document.createElement("h1");
			document.body.append(heading1);

			render(<Header {...defaultProps} skipLinkId="super-skip-link" />);
			expect(heading1).toHaveAttribute("id", "super-skip-link");
			heading1.parentNode.removeChild(heading1);
		});
	});

	describe("Dropdown open and close callbacks", () => {
		it("should call onDropdownOpen function name callback prop if dropdown is opened", async () => {
			window.testFunc = jest.fn();

			const { getByRole } = render(
					<Header {...defaultProps} onDropdownOpen={"testFunc"} />
				),
				user = userEvent.setup(),
				guidanceMenuButton = getByRole("button", { name: "Guidance" });

			await user.click(guidanceMenuButton);

			expect(window.testFunc).toHaveBeenCalledTimes(1);
		});

		it("should call onDropdownClose function name callback prop if dropdown is opened", async () => {
			window.onDropdownCloseFunc = jest.fn();

			const { getByRole } = render(
					<Header {...defaultProps} onDropdownClose={"onDropdownCloseFunc"} />
				),
				user = userEvent.setup(),
				guidanceMenuButton = getByRole("button", { name: "Guidance" });

			await user.click(guidanceMenuButton);
			await user.click(guidanceMenuButton);

			expect(window.onDropdownCloseFunc).toHaveBeenCalledTimes(2);
		});

		it("should call onDropdownOpen callback prop if dropdown is opened", async () => {
			const onDropdownOpen = jest.fn(),
				{ getByRole } = render(
					<Header {...defaultProps} onDropdownOpen={onDropdownOpen} />
				),
				user = userEvent.setup(),
				guidanceMenuButton = getByRole("button", { name: "Guidance" });

			await user.click(guidanceMenuButton);

			expect(onDropdownOpen).toHaveBeenCalledTimes(1);
		});

		it("should call onDropdownClose callback prop if dropdown is closed", async () => {
			const onDropdownClose = jest.fn(),
				{ getByRole } = render(
					<Header {...defaultProps} onDropdownClose={onDropdownClose} />
				),
				user = userEvent.setup(),
				guidanceMenuButton = getByRole("button", { name: "Guidance" });

			await user.click(guidanceMenuButton);
			await user.click(guidanceMenuButton);

			expect(onDropdownClose).toHaveBeenCalledTimes(2);
		});
	});

	describe("Logo tracking", () => {
		it("should track logo click", async () => {
			const { getByRole } = render(<Header {...defaultProps} />),
				user = userEvent.setup(),
				homeLink = getByRole("link", { name: "Home" });

			await user.click(homeLink);

			expect(window.dataLayer).toEqual([
				{
					destinationUrl: "https://www.nice.org.uk/",
					event: eventName,
					eventCategory: defaultEventCategory,
					eventAction: headerClickEventAction,
					eventLabel: "Logo",
					eventCallback: expect.any(Function),
					eventTimeout: eventTimeout,
				},
			]);
		});

		it("should prevent default and navigate in event callback on logo click", () => {
			const { getByRole } = render(<Header {...defaultProps} />),
				homeLink = getByRole("link", { name: "Home" }),
				clickEvent = createEvent.click(homeLink);

			fireEvent(homeLink, clickEvent);

			expect(clickEvent.defaultPrevented).toBe(true);

			window.dataLayer[0].eventCallback();
			expect(window.location.href).toEqual("https://www.nice.org.uk/");
		});
	});

	describe("Sign in button", () => {
		it("Should not render sign in button if false is supplied to auth prop", () => {
			const { getByRole, rerender } = render(<Header {...defaultProps} auth />),
				signInButton = getByRole("link", { name: "Sign in" });

			expect(signInButton).toBeInTheDocument();

			rerender(<Header {...defaultProps} auth={false} />);

			expect(signInButton).not.toBeInTheDocument();
		});
	});
});
