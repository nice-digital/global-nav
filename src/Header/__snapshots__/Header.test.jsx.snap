// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Matches snapshot 1`] = `
<HeaderContextProvider>
  <ContextConsumer>
    <Component />
  </ContextConsumer>
  <div
    className="header"
    data-tracking="Global nav"
    id="top"
  >
    <header
      aria-label="Site header"
    >
      <ul
        aria-label="Accessibility links"
        className="a11yLinks"
      >
        <li>
          <SkipLink
            to="#content-start"
          >
            Skip to content
          </SkipLink>
        </li>
        <li>
          <SkipLink
            to="https://www.nice.org.uk/accessibility"
          >
            Accessibility help
          </SkipLink>
        </li>
      </ul>
      <div
        className="container"
        id="site-header"
      >
        <a
          aria-label="Home"
          className="home"
          data-tracking="Logo"
          href="https://www.nice.org.uk/"
          onClick={[Function]}
        >
          <SvgLogoFull
            className="icon"
            height={null}
            width={null}
          />
        </a>
        <div
          className="wrapper"
        >
          <div
            className="search"
          >
            <Search
              onNavigating="onNavigatingHandler"
              placeholder="Search NICEâ€¦"
              query=""
              skipLinkId="content-start"
              url="/search"
            />
          </div>
          <button
            aria-controls="header-menu"
            aria-expanded={false}
            aria-haspopup="menu"
            aria-label="Expand site menu"
            className="mobileMenuBtn"
            id="header-menu-button"
            onClick={[Function]}
            type="button"
          >
            Menu
          </button>
          <div
            className="account"
          >
            <Account
              accountsData={null}
              environment="live"
              isLoggedIn={false}
              onLoginStatusChecked={[Function]}
              provider="niceAccounts"
            />
          </div>
        </div>
      </div>
      <Nav
        accountsLinks={null}
        additionalSubMenuItems={Array []}
        isExpanded={false}
        onNavigating="onNavigatingHandler"
        service={null}
        skipLinkId="content-start"
      />
    </header>
    <CoronaMessage />
    <OldIEMessage />
  </div>
</HeaderContextProvider>
`;
